control/discovery.py

```diff
@@ -0,0 +1,1127 @@
+        # RTD3R, RTD3E
+        ("todo_fields3", c_ubyte * 8),
+        # optional asynchronous events supported
+        ("oaes", c_ubyte * 4),
+        # CTRATT, RRLS, CNTRLTYPE, FGUID, NVMe Management Interface, OACS, ACL
+        ("todo_fields4", c_ubyte * 163),
+        # asynchronous events request limit
+        ("aerl", c_uint8),
+        ("firmware_updates", c_uint8),
+        # log page attributes
+        ("lpa", c_uint8),
+        # error log page entries(ELPE)
+        ("elpe", c_uint8),
+        # NPSS, AVSCC, APSTA, WCTEMP, CCTEMP, MTFA, HMPRE, HMIN, TNVMCAP...
+        # TODO: keep alive support - timer value(KAS)?
+        ("todo_fields5", c_ubyte * 251),
+        # maximum outstanding commands
+        ("max_cmd", c_uint16),
+        # number of namespace, optional NVM command support
+        ("todo_fields6", c_uint8 * 6),
+        # fused operation support
+        ("fused_operation", c_uint16),
+        # FNA, VWC, AWUN, AWUPF, NVSCC, NWPC
+        ("todo_fields7", c_uint8 * 8),
+        # atomic compare & write unit
+        ("acwu", c_uint16),
+        ("reserved1", c_uint16),
+        # SGL support
+        ("sgls", c_uint8 * 4),
+        # maxinum number of allowed namespaces
+        ("mnan", c_uint32),
+        ("reserved2", c_ubyte * 224),
+        ("subnqn", c_ubyte * 256),
+        ("reserved3", c_ubyte * 768),
+        ("nvmeof_attributes", c_ubyte * 256),
+        ("power_state_attributes", c_ubyte * 1024),
+        ("vendor_specific", c_ubyte * 1024)
+    ]
+
+    def compose_reply(self):
+        return bytes(self)
+
+# for set feature, keep alive and async
+class  CqeNVMe(Structure):
+    _fields_ = [
+        ("dword0", c_uint32),
+        ("dword1", c_uint32),
+        ("sq_head_ptr", c_uint16),
+        ("sq_id", c_uint16),
+        ("cmd_id", c_uint16),
+        ("status", c_uint16)
+    ]
+
+    def compose_reply(self):
+        return bytes(self)
+
+class NVMeGetLogPage(Structure):
+    _fields_ = [
+        # generation counter
+        ("genctr", c_uint64),
+        # number of records
+        ("numrec", c_uint64),
+        #record format
+        ("recfmt", c_uint16),
+        ("reserved", c_ubyte * 1006)
+    ]
+
+    def compose_short_reply(self):
+        return bytes(self)[:16]
+
+    def compose_data_reply(self):
+        return bytes(self)
+
+class DiscoveryLogEntry(Structure):
+    _fields_ = [
+        ("trtype", c_uint8),
+        ("adrfam", c_uint8),
+        ("subtype", c_uint8),
+        ("treq", c_uint8),
+        ("port_id", c_uint16),
+        ("controller_id", c_uint16),
+        # admin max SQ size
+        ("asqsz", c_uint16),
+        ("reserved1", c_ubyte * 22),
+        ("trsvcid", c_ubyte * 32),
+        ("reserved2", c_ubyte * 192),
+        ("subnqn", c_ubyte * 256),
+        ("traddr", c_ubyte * 256),
+        # Transport specific address subtype
+        ("tsas", c_ubyte * 256)
+    ]
+
+    def compose_reply(self):
+        return bytes(self)
```

Comment:
Let's leverage Python syntactic sugar:
```suggestion
class AutoSerializableStructure(Structure):
    def __add__(self, other):
        if isinstance(other, Structure):
            return bytes(self) + bytes(other)
        else:
            raise ValueError("...")
            
class Pdu(AutoSerializableStructure):
    ...
```
This way we can do:
```py
conn.sendall(pdu_reply + keep_alive_reply)
```
BTW, this will involve creating and copying several buffers, so if performance becomes an issue for this, here we could easily replace this with [a preallocated buffer](https://docs.python.org/3/library/ctypes.html#ctypes.create_string_buffer), or given we're using an streaming socket, I guess that we can send each sub-buffer independently:
```py
conn.send(pdu_reply)
conn.send(keep_alive_reply)
```

Reply:
Thank you for your guidance. I have learned, and there are two special statements. I have added a `bytes` type method to the `AutoSerializableStructure` class. Please help take a look:
```
if isinstance(other, LittleEndianStructure):
            return bytes(self) + bytes(other)
        elif isinstance(other, bytes):
            return bytes(self) + other
```

line 375: `reply = pdu_reply + nvme_tcp_data_pdu + bytes(nvme_get_log_page_reply)[:16]`
line 790: `conn.sendall(pdu_reply + icresp_reply + bytes(112))`